File: 0742-closest-leaf-in-a-binary-tree/0742-closest-leaf-in-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode dfs(TreeNode node,HashMap<Integer,TreeNode>bm,int target) {
        if(node==null) return null;
        if(target==node.val) {
            return node;
        }
        TreeNode left = dfs(node.left,bm,target);
        if(left != null) {
            bm.put(node.left.val,node);
            return left;
        }
        TreeNode right = dfs(node.right,bm,target);
        if(right != null) {
            bm.put(node.right.val,node);
            return right;
        }
        return null;
    }
    public int findClosestLeaf(TreeNode root, int k) {
        if(root.left==null && root.right==null) return root.val;
        HashMap<Integer,TreeNode> bm = new HashMap<Integer,TreeNode>();
        LinkedList<TreeNode> que = new LinkedList<>();
        HashSet<Integer> vis = new HashSet<>();
        TreeNode start = dfs(root,bm,k);
        que.addLast(start);
        System.out.print(start.val);
        while(que.size()!=0) {
            int size = que.size();
            while(size-->0) {
                TreeNode rn = que.removeFirst();
                if(rn.left == null && rn.right == null) {
                    return rn.val;
                }
                if(rn.left!=null && vis.add(rn.left.val)) {
                    que.addLast(rn.left);
                    
                }
                if(rn.right!=null && vis.add(rn.right.val)) {
                    que.addLast(rn.right);
                }
                
                if(bm.containsKey(rn.val) && vis.add(bm.get(rn.val).val)) {
                    que.addLast(bm.get(rn.val));
                }
            }
        }
       return -1; 
    }
}

File: 0742-closest-leaf-in-a-binary-tree/0742-closest-leaf-in-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* getTar(TreeNode* node,unordered_map<TreeNode*,TreeNode*>& map,int target) {
        if(!node) return NULL;
        if(node->val == target) return node;
        
        TreeNode* left = getTar(node->left,map,target);
        if(left) {
            map[node->left] = node;
            return left;
        }
        TreeNode* right = getTar(node->right,map,target);
        if(right) {
            map[node->right] = node;
            return right;
        }
        return NULL;
    }
    int findClosestLeaf(TreeNode* root, int k) {
        unordered_map<TreeNode*,TreeNode*> map;
        TreeNode* tar = getTar(root,map,k);
        set<int> vis;
        queue<TreeNode*> que;
        que.push(tar);
        while(que.size()!=0) {
            int size = que.size();
            while(size-->0) {
                TreeNode* rn = que.front(); que.pop();
                if(!rn->left && !rn->right) {
                    return rn->val;
                }
                if(rn->left && vis.find(rn->left->val) == vis.end()) {
                    que.push(rn->left);
                    vis.insert(rn->left->val);
                }
                
                if(rn->right && vis.find(rn->right->val)==vis.end()) {
                    que.push(rn->right);
                    vis.insert(rn->right->val);
                }
                
                if(map.find(rn) != map.end() && vis.find(map[rn]->val)==vis.end()) {
                    que.push(map[rn]);
                    vis.insert(rn->val);
                }
            }
        }
        return -1;
    }
};